### ISSUE TITLE
The old symbolic links are not deleted when the NVMe partition is renamed

### ISSUE THREAD

**Issue by chenglongshe** [systemd/systemd#27981]
### systemd version the issue has been seen with

systemd 253 (253-2728-g159f1b7)

### Used distribution

FusionOS-23_23.0.2.B002_x86_64

### Linux kernel version used

5.10.0-136.30.0.106.u42.fos23.x86_64

### CPU architectures issue was seen on

x86_64

### Component

systemd-udevd

### Expected behaviour you didn't see

ls -l  /dev/disk/by-partlabel/ |grep xxxx
lrwxrwxrwx 1 root root 15 Jun  9 08:19 xxxxx200 -> ../../nvme0n1p3

### Unexpected behaviour you saw

lrwxrwxrwx 1 root root 15 Jun  9 08:29  xxxxx106 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:29  xxxxx120 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:30  xxxxx125 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:30  xxxxx127 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:26  xxxxx13 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:30  xxxxx131 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:26  xxxxx14 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:30  xxxxx146 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:26  xxxxx15 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:31  xxxxx150 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:31  xxxxx153 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:31  xxxxx156 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:31  xxxxx158 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:31  xxxxx159 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:31  xxxxx160 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:31  xxxxx169 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:31  xxxxx173 -> ../../nvme0n1p3

### Steps to reproduce the problem

prepare: (nvme disk: create 40 partitions)
parted /dev/nvme0n1 --script mkpart primary1 ext4 103G 105G
parted /dev/nvme0n1 --script mkpart primary2 ext4 106G 108G
parted /dev/nvme0n1 --script mkpart primary3 ext4 109G 111G
parted /dev/nvme0n1 --script mkpart primary4 ext4 112G 114G
parted /dev/nvme0n1 --script mkpart primary5 ext4 115G 117G
...
parted /dev/nvme0n1 --script mkpart primary32 ext4 196G 198G
parted /dev/nvme0n1 --script mkpart primary33 ext4 199G 201G
parted /dev/nvme0n1 --script mkpart primary34 ext4 202G 204G
parted /dev/nvme0n1 --script mkpart primary35 ext4 205G 207G
parted /dev/nvme0n1 --script mkpart primary36 ext4 208G 210G
parted /dev/nvme0n1 --script mkpart primary37 ext4 211G 213G
parted /dev/nvme0n1 --script mkpart primary38 ext4 214G 216G
parted /dev/nvme0n1 --script mkpart primary39 ext4 217G 219G
parted /dev/nvme0n1 --script mkpart primary40 ext4 220G 222G

test1.sh: (change label)
for i in `seq 200`; do
parted /dev/nvme0n1 --script name 3 xxxxx$i;
sleep 2
done

test2.sh: (repeat add, remove)
for i in `seq 200`; do
parted /dev/nvme0n1 --script mkpart primary32 ext4 196G 198G
sleep 2
parted /dev/nvme0n1 --script print  |grep -w 198GB | awk '{print $1}' | xargs parted /dev/nvme0n1 --script rm
done

test3.sh:  (repeat add, remove)
for i in `seq 200`; do
parted /dev/nvme0n1 --script mkpart primary202 ext4 202G 204G
sleep 2
parted /dev/nvme0n1 --script print  |grep -w 204GB | awk '{print $1}' | xargs parted /dev/nvme0n1 --script rm
done

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
KERNEL[49574.218161] change   /devices/pci0000:e2/0000:e2:02.0/0000:e3:00.0/nvme/nvme0/nvme0n1/nvme0n1p14 (block)
ACTION=change
MAJOR=259
MINOR=45             
in processing change event:
device_clone_with_db() {
	device_shallow_clone() -> device_set_devnum(dest, major=259, minor=45)
	device_read_uevent_file() --> change major:minor
	device_read_db_internal() --> read WRONG udev database file:run/udev/data/b259:15 (device_get_device_id RETURN b259:15) // introduce problem points
}

KERNEL[49574.318800] remove   /devices/pci0000:e2/0000:e2:02.0/0000:e3:00.0/nvme/nvme0/nvme0n1/nvme0n1p14 (block)
ACTION=remove
MAJOR=259
MINOR=45

KERNEL[49574.320786] add      /devices/pci0000:e2/0000:e2:02.0/0000:e3:00.0/nvme/nvme0/nvme0n1/nvme0n1p14 (block)
ACTION=add
MAJOR=259
MINOR=15


KERNEL[49576.297689] change   /devices/pci0000:e2/0000:e2:02.0/0000:e3:00.0/nvme/nvme0/nvme0n1/nvme0n1p14 (block)
ACTION=change
MAJOR=259
MINOR=15
```

**Comment by chenglongshe** [comment]
I think the commit that introduced the problem was  https://github.com/systemd/systemd/pull/19815/commits/381f6d4ba5551898e7ff19189485072b94879281，
when udevd handles the change event and the remove event occurs at the same time, in the function device_shallow_clone, the content of the uevent file may have changed. If it is not necessary, it is best not to call device_read_uevent_file to read this file

**Comment by chenglongshe** [comment]
the following modification was made in the device_shallow_clone function, and the problem of remaining symbolic links was not found after verification

--- a/src/libsystemd/sd-device/device-private.c
+++ b/src/libsystemd/sd-device/device-private.c
@@ -786,11 +786,20 @@ int device_shallow_clone(sd_device *old_device, sd_device **new_device) {

         if (sd_device_get_property_value(old_device, "MAJOR", &val) >= 0) {
                 const char *minor = NULL;
+                const char *id = NULL;

                 (void) sd_device_get_property_value(old_device, "MINOR", &minor);
+
                 r = device_set_devnum(ret, val, minor);
                 if (r < 0)
                         return r;
+
+                r = device_get_device_id(old_device, &id)
+                if (r >= 0) {
+                        /* copy device_id from old_device but uevent file. */
+                        // log_device_info(old_device, "device_shallow_clone: set device_id = [%s]", id);
+                        ret->device_id = strdup(id);
+                }
         }

         /* And then read uevent file, but ignore errors, as some devices seem to return a spuriou

**Comment by yuwata** [comment]
Thank you for the report. I am working on the issue.

**Comment by yuwata** [comment]
Fix is waiting in #28255.
